### PR TITLE
New Plugin: MutualGroupDMs

### PR DESCRIPTION
**Pull request by aamiaa**
This plugin adds a Mutual Group DMs tab on user profile, which lets you view all group dms you share with the user and navigate to them.

### CODE REVIEW

File: src/plugins/mutualGroupDMs.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    description: "Shows mutual group dms on profile.",
+    authors: [Devs.amia],
+
+    patches: [
+        {
+            find: ".Messages.USER_PROFILE_MODAL", // Note: the module is lazy-loaded
+            replacement: [
+                {
+                    match: /(?<=(\i)\.isClyde\(\)\?null:(\(0,\i\.\i\).+?id:)\i\.\i\.MUTUAL_GUILDS(,children:)\i\.\i\.Messages\.MUTUAL_GUILDS}\))/,
+                    replace: ",!$1.bot?$2\"MUTUAL_GDMS\"$3\"Mutual GDMs\"}):null" // The label is "Mutual GDMs" because otherwise it's too long when Activity tab is present
```
Comment (TheKodeToad) [inline src/plugins/mutualGroupDMs.tsx:58]: As a muggle i just call them groups

Comment (aamiaa) [reply src/plugins/mutualGroupDMs.tsx:58]: I labeled it "Mutual Group DMs" before, but "Mutual Groups" seems to work

File: src/plugins/mutualGroupDMs.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+const ProfileListClasses = findByPropsLazy("emptyIconFriends", "emptyIconGuilds");
+const GuildLabelClasses = findByPropsLazy("guildNick", "guildAvatarWithoutIcon");
+
+// @ts-ignore
+const ScrollerThin = LazyComponent(() => Menu.ScrollerThin);
+// @ts-ignore
+const Clickable = LazyComponent(() => Menu.Clickable);
+// @ts-ignore
+const Avatar = LazyComponent(() => Menu.Avatar);
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:39]: don't do this, you're not supposed to use anything from Menu that isn't a Menu. It's just the entire module for simplicity. You will instead want to add additional members to this line:
https://github.com/Vendicated/Vencord/blob/3e3d05fc26a634b17549c9473bd8aeebb7dec4ec/src/webpack/common/components.ts#L57

File: src/plugins/mutualGroupDMs.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                {
+                    match: /({user:(\i),onClose:(\i)}\);)(case \i\.\i\.MUTUAL_FRIENDS)/,
+                    replace: "$1case \"MUTUAL_GDMS\":return $self.renderMutualGDMs($2,$3);$4"
+                }
+            ]
+        }
+    ],
+
+    renderMutualGDMs(user: User, onClose: Function) {
+        const entries = ChannelStore.getSortedPrivateChannels().filter(x => x.type === 3 && x.recipients.includes(user.id)).map(x => {
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:69]: magic numbers are bad, for better readability please extract it into a properly named constant

File: src/plugins/mutualGroupDMs.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            return <Clickable className={ProfileListClasses.listRow} onClick={() => {
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:70]: it's prettier if u change it from
```js
.map(x => {
	return <Clickable ...
})
```
to
```js
.map(x => (
	<Clickable ...
))
```

File: src/plugins/mutualGroupDMs.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+const GuildLabelClasses = findByPropsLazy("guildNick", "guildAvatarWithoutIcon");
+
+// @ts-ignore
+const ScrollerThin = LazyComponent(() => Menu.ScrollerThin);
+// @ts-ignore
+const Clickable = LazyComponent(() => Menu.Clickable);
+// @ts-ignore
+const Avatar = LazyComponent(() => Menu.Avatar);
+
+function GetGroupDMName(channel: Channel) {
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:41]:
```suggestion
function getGroupDMName(channel: Channel) {
```

File: src/plugins/mutualGroupDMs.tsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    return channel.name || channel.recipients.map(UserStore.getUser).filter(x => x != null).map(x => {
+        return RelationshipStore.getNickname(x.id) || UserUtils.getName(x);
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:43]: same as below

File: src/plugins/mutualGroupDMs.tsx (hunk 2 above)
Comment (aamiaa) [reply src/plugins/mutualGroupDMs.tsx]: okay I did my best to type those components, idk if I got everything correctly though

File: src/plugins/mutualGroupDMs.tsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    patches: [
+        {
+            find: ".Messages.USER_PROFILE_MODAL", // Note: the module is lazy-loaded
+            replacement: [
+                {
+                    match: /(?<=(\i)\.isClyde\(\)\?null:(\(0,\i\.\i\).+?id:)\i\.\i\.MUTUAL_GUILDS(,children:)\i\.\i\.Messages\.MUTUAL_GUILDS}\))/,
+                    replace: ",!$1.bot?$2\"MUTUAL_GDMS\"$3\"Mutual Groups\"}):null"
+                },
+                {
+                    match: /({user:(\i),onClose:(\i)}\);)(case \i\.\i\.MUTUAL_FRIENDS)/,
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:53]: no need to match either of these
```suggestion
                    match: /(?<={user:(\i),onClose:(\i)}\);)(?=case \i\.\i\.MUTUAL_FRIENDS)/,
```

File: src/plugins/mutualGroupDMs.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                    replace: "$1case \"MUTUAL_GDMS\":return $self.renderMutualGDMs($2,$3);$4"
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:54]:
```suggestion
                    replace: "case \"MUTUAL_GDMS\":return $self.renderMutualGDMs($1,$2);"
```

File: src/plugins/mutualGroupDMs.tsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    name: "MutualGroupDMs",
+    description: "Shows mutual group dms on profile.",
+    authors: [Devs.amia],
+
+    patches: [
+        {
+            find: ".Messages.USER_PROFILE_MODAL", // Note: the module is lazy-loaded
+            replacement: [
+                {
+                    match: /(?<=(\i)\.isClyde\(\)\?null:(\(0,\i\.\i\).+?id:)\i\.\i\.MUTUAL_GUILDS(,children:)\i\.\i\.Messages\.MUTUAL_GUILDS}\))/,
```
Comment (Vendicated) [inline src/plugins/mutualGroupDMs.tsx:49]: this looks pretty bad. `.+` is really dangerous (even with ?) and you should avoid it if possible

**Review by Vendicated — CHANGES_REQUESTED** [review]

**Review by Vendicated — APPROVED** [review]

**Comment by Vendicated** [comment]
thank you, pretty cool!!

**Comment by D3SOX** [comment]
Would be nice if this was also shown on the profile sidebar :)
image

**Pull request merged**